internal/cmd/root/root.go

```
@@ -129,36 +130,45 @@ func Execute() {
+		// since we do not always want to print a stack trace even in verbose
+		// mode
+		icon := "❌ "
+		style := pterm.Style{pterm.Bold, pterm.FgRed}
+		stackTraceStyle := pterm.Style{pterm.Fuzzy}
+		if log.PlainStyle() {
+			icon = ""
+			style = pterm.Style{}
+			stackTraceStyle = pterm.Style{}
+		}
+
+		// Error types we want to handle differently
 		var usageErr *cmdutils.IncorrectUsageError
+		var couldBeSandboxError *cmdutils.CouldBeSandboxError
+		var signalErr *cmdutils.SignalError
+		var execErr *cmdutils.ExecError
+		var apiErr *api.APIError
+		var silentErr *cmdutils.SilentError
+
 		if errors.As(err, &usageErr) ||
 			strings.HasPrefix(err.Error(), "unknown command") ||
 			regexp.MustCompile(`(accepts|requires).*arg\(s\)`).MatchString(err.Error()) {
-
-			// Ensure that there is an extra newline between the error
-			// and the usage message
-			if !strings.HasSuffix(err.Error(), "\n") {
-				_, _ = fmt.Fprintln(cmd.ErrOrStderr())
-			}
-
 			// Make cmd.Help() print to stderr
 			cmd.SetOut(cmd.ErrOrStderr())
+
 			// Print the usage message of the command. We use cmd.Help()
 			// here instead of cmd.UsageString() because the latter
 			// doesn't include the long description.
 			_ = cmd.Help()
+
+			// Add "Invalid Usage" in front of error message
+			err = errors.Wrap(err, "Invalid usage")
 		}
 
-		var couldBeSandboxError *cmdutils.CouldBeSandboxError
 		if errors.As(err, &couldBeSandboxError) {
-			// Ensure that there is an extra newline between the error
-			// and the following message
-			if !strings.HasSuffix(err.Error(), "\n") {
-				_, _ = fmt.Fprintln(cmd.ErrOrStderr())
-			}
-			msg := `Note: If you don't expect this fuzz test to do any harm to the system
+			msg := `
+Note: This error could have occurred due to sandboxing.
+If you don't expect this fuzz test to do any harm to the system
 accidentally (like overwriting files), you might want to try
 running it without sandboxing:
```

We want to print this note after the error message, so this if block should be moved below the one which prints non-silent errors.

It (accidentally) worked as expected before, because we handled all errors which we wrapped as `CouldBeSandboxError` on an upper layer and returned a `SilentError`.

---

While testing this, I noticed that we didn't print the note for the `CouldBeSandboxError` returned here: https://github.com/CodeIntelligenceTesting/cifuzz/blob/mlsandnerCI%2FCLI-1189-error-no-stacktrace/internal/cmd/coverage/llvm/llvm.go#L89-L89

I fixed that in a commit which I pushed to this branch, @mlsandnerCI please review that commit.

---

Can a signal error also be a sandbox error? Because then I would need to remove the os.Exit from that if-clause so that we can still print this error message if it is checked after the non-silent error check 🤔

---

No, we return a `SignalError` when cifuzz catches a terminating signal and a `CouldBeSandboxError` when a command that's executed in Minijail fails. But it's good that you considered that case 👌